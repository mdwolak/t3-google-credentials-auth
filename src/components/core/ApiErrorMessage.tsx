import { useState } from "react";

import { Alert } from "~/components/core";

export const ApiErrorMessage = ({
  error,
  visible = true,
  header = "Oops!",
}: {
  error: { message: string } | null;
  visible?: boolean;
  header?: string;
}) => {
  const [expand, setExpand] = useState(false);

  if (error) {
    console.log(error.message);
    if (visible)
      return (
        <Alert severity="error">
          {header && <strong className="font-bold">{header} </strong>}
          {error.message}
          {location.hostname === "localhost" && (
            <div>
              <a
                className="mb-2 block cursor-pointer text-indigo-600 hover:text-indigo-500 hover:underline"
                onClick={() => setExpand(!expand)}>
                {expand ? "Collapse" : "Expand"}
              </a>
              {expand && <pre>Error: {JSON.stringify(error, null, 2)}</pre>}
            </div>
          )}
        </Alert>
      );
  }

  return null;
};
