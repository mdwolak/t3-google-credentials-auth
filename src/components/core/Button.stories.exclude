/**
 * 
 *               <Button variant="primary">Sign in</Button>
              <Button variant="primary" disabled>
                Sign in
              </Button>
              <Button variant="primary" isLoading={true}>
                Sign in
              </Button>

              <Button variant="secondary">Sign in</Button>
              <Button variant="secondary" disabled>
                Sign in
              </Button>
              <Button variant="secondary" isLoading={true}>
                Sign in
              </Button>

              <Button variant="destructive">Sign in</Button>
              <Button variant="destructive" disabled>
                Sign in
              </Button>
              <Button variant="destructive" isLoading={true}>
                Sign in
              </Button>

 */
