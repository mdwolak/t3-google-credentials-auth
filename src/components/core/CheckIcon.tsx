function CheckIcon(props) {
  return (
    <svg
      viewBox="0 0 20 20"
      aria-hidden="true"
      className="h-5 w-5 flex-none fill-emerald-500 stroke-white dark:fill-emerald-200/20 dark:stroke-emerald-200"
      {...props}>
      <circle cx="10" cy="10" r="10" strokeWidth="0" />
      <path
        fill="none"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="1.5"
        d="m6.75 10.813 2.438 2.437c1.218-4.469 4.062-6.5 4.062-6.5"
      />
    </svg>
  );
}
