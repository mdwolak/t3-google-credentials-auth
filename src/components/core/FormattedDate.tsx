import React from "react";

import { format as fnsFormat } from "date-fns";

type FormattedDateProps = {
  date: Date;
  format?: string;
};

export const FormattedDate = ({ date, format = "PPP" }: FormattedDateProps) => {
  const formattedDate = fnsFormat(date, format);

  return <span className="date">{formattedDate}</span>;
};
