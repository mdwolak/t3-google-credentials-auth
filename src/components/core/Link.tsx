import NextLink, { type LinkProps as NextLinkProps } from "next/link";
import type { ComponentPropsWithoutRef } from "react";

import classNames from "classnames";

const variants = {
  primary: "text-indigo-600 hover:text-indigo-900 font-medium hover:underline",
  secondary: "text-gray-600 hover:text-gray-900 hover:underline",
};

export type LinkProps = Omit<ComponentPropsWithoutRef<"a">, "href"> &
  NextLinkProps & {
    variant?: keyof typeof variants;
  };

export const Link = (props: LinkProps) => {
  const {
    href,
    variant = "primary",
    as,
    replace,
    scroll,
    shallow,
    prefetch,
    locale,
    className,
    children,
    ...rest
  } = props;

  return (
    <NextLink
      passHref={true}
      href={href}
      as={as}
      replace={replace}
      scroll={scroll}
      shallow={shallow}
      prefetch={prefetch}
      locale={locale}
      className={classNames(variants[variant], className)}
      {...rest}>
      {children}
    </NextLink>
  );
};
