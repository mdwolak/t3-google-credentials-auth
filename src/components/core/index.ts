export * from "react-hot-toast"; //TODO: Remove this

export * from "./Alert";
export * from "./ApiErrorMessage";
export * from "./Button";
export * from "./FormattedDate";
export * from "./Link";
export * from "./SEOHead";
