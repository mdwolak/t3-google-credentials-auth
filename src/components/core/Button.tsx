import type { ComponentPropsWithRef, ComponentPropsWithoutRef } from "react";
import { type Ref, forwardRef } from "react";

import classNames from "classnames";

import { Spinner } from "~/components/core/Spinner";

const variants = {
  primary:
    "bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500",
  secondary: "bg-white text-gray-900 ring-1 ring-inset ring-gray-300 enabled:hover:bg-gray-50",
  destructive: "bg-red-600 text-white border-transparent enabled:hover:opacity-80",
};

/**
 * Size-sensitive classes
 */
const textWithOptionalIconClasses = {
  xs: "px-2 py-1 text-xs", //fits compact table, inline text, bulleted lists
  sm: "px-2.5 py-1.5 text-sm", //fits standard table
  default: "px-3 py-2 text-sm", //default action button
  lg: "px-4 py-3 text-md", //big conspicuous button
};

const iconOnlyClasses = {
  xs: "p-1",
  sm: "p-1.5",
  default: "p-2",
  lg: "p-3.5",
};

const iconSizes = {
  xs: "h-2.5 w-2.5",
  sm: "h-4 w-4",
  default: "h-4 w-4",
  lg: "h-5 w-5",
};

type SVGComponent = React.FunctionComponent<React.SVGProps<SVGSVGElement>>;

type ButtonProps = ComponentPropsWithRef<"button"> & {
  isLoading?: boolean;
  variant?: keyof typeof variants;
  Icon?: SVGComponent | React.ElementType; //accept lucid-react icons | heroicons
  size?: keyof typeof textWithOptionalIconClasses;
  fullWidth?: boolean;
  rounded?: boolean;
};

export const Button = forwardRef(function Button(
  {
    children,
    className,
    type = "button",
    isLoading,
    disabled,
    variant = "primary",
    Icon,
    size = "default",
    fullWidth = false,
    rounded = false,
    ...props
  }: ButtonProps,
  ref: Ref<HTMLButtonElement>
) {
  disabled = disabled || isLoading;
  const iconClass = `${children ? "mr-2 " : ""}${iconSizes[size]}`;

  return (
    <button
      ref={ref}
      type={type}
      className={classNames(
        fullWidth && "flex w-full justify-center",
        (Icon || isLoading) && "inline-flex items-center", //items-center centers icon horizontally
        rounded ? "rounded-full" : "rounded-md",
        "font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60", //font-semibold border
        children ? textWithOptionalIconClasses[size] : iconOnlyClasses[size],
        variants[variant],
        className
      )}
      disabled={disabled}
      {...props}>
      {isLoading ? (
        <>
          <Spinner className={iconClass} aria-hidden="true" />
          <span>Processing...</span>
        </>
      ) : (
        <>
          {/* <span >{icon}</span> */}
          {Icon && <Icon className={iconClass} aria-hidden="true" />}
          {children}
        </>
      )}
    </button>
  );
});

type IconButtonProps = ComponentPropsWithoutRef<"button"> & {
  Icon: SVGComponent | React.ElementType;
  srText: string;
  className?: string;
};
export const IconButton = ({ Icon, srText, className, ...props }: IconButtonProps) => {
  return (
    <button
      type="button"
      className={classNames("text-gray-400 hover:text-gray-500", className)}
      {...props}>
      <Icon className="h-6 w-6" aria-hidden="true" />
      <span className="sr-only">{srText}</span>
    </button>
  );
};

export default IconButton;
