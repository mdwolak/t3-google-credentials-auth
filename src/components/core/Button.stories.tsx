import type { Meta, StoryObj } from "@storybook/react";

import { Button } from "./Button";

const meta = {
  title: "Core/Button",
  component: Button,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {
    //backgroundColor: { control: "color" },
  },
  args: {
    children: "Button",
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

/*
 * STORIES
 */
export const Primary: Story = {
  args: {
    variant: "primary",
  },
};

export const Primary_Loading: Story = {
  args: {
    ...Primary.args,
    isLoading: true,
  },
};

export const Secondary: Story = {
  args: {
    variant: "secondary",
  },
};

export const Secondary_Loading: Story = {
  args: {
    variant: "secondary",
    isLoading: true,
  },
};

export const Secondary_Disabled: Story = {
  args: {
    variant: "secondary",
    disabled: true,
  },
};

export const Destructive: Story = {
  args: {
    variant: "destructive",
  },
};

export const Destructive_Loading: Story = {
  args: {
    variant: "destructive",
    isLoading: true,
  },
};

export const Destructive_Disabled: Story = {
  args: {
    variant: "destructive",
    disabled: true,
  },
};
