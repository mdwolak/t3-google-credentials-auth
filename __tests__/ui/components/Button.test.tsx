/**
 * @jest-environment jsdom
 */
import { PlusIcon, XMarkIcon } from "@heroicons/react/24/outline";
import { render, screen } from "@testing-library/react";

import { Button } from "~/components/core";

describe("Button", () => {
  describe("Primary", () => {
    it("text", () => {
      render(<Button>Click</Button>);
      const button = screen.getByRole("button", { name: "Click" });
      expect(button).toBeInTheDocument();
      expect(button).toMatchSnapshot();
    });

    it("text and icon", () => {
      render(<Button Icon={PlusIcon}>Click</Button>);
      const button = screen.getByRole("button", { name: "Click" });
      expect(button).toBeInTheDocument();
      expect(button).toMatchSnapshot();

      const icon = document.querySelector("svg") as SVGSVGElement;
      expect(button).toContainElement(icon);
      expect(icon.classList.toString()).toMatchInlineSnapshot(`"mr-2 h-5 w-5"`);
    });

    it("icon only", () => {
      render(<Button Icon={PlusIcon} />);
      const button = screen.getByRole("button");

      expect(button).toBeInTheDocument();
      expect(button).toMatchSnapshot();

      const icon = document.querySelector("svg") as SVGSVGElement;
      expect(button).toContainElement(icon);
      expect(icon.classList.toString()).toEqual("h-5 w-5");
    });

    it("loading", () => {
      render(<Button isLoading={true}>Click</Button>);
      const button = screen.getByRole("button");

      expect(button).toBeInTheDocument();
      expect(button).toMatchSnapshot();

      const spinner = screen.getByTestId("spinner");
      expect(button).toContainElement(spinner);
      expect(spinner).toHaveClass("animate-spin mr-2 h-5 w-5", { exact: true });
    });

    it("disabled", () => {
      render(<Button disabled>Click</Button>);
      const button = screen.getByRole("button");

      expect(button).toBeInTheDocument();
      expect(button).toMatchSnapshot();

      const icon = document.querySelector("svg") as SVGSVGElement;
      expect(icon).toBeNull();
    });
  });
});

describe("secondary", () => {
  it("text", () => {
    render(<Button>Click</Button>);
    const button = screen.getByRole("button", { name: "Click" });
    expect(button).toBeInTheDocument();
    expect(button).toMatchSnapshot();
  });

  it("rounded", () => {
    render(<Button rounded>Click</Button>);
    const button = screen.getByRole("button", { name: "Click" });
    expect(button).toBeInTheDocument();
    expect(button).toMatchSnapshot();
  });

  it("fullWidth", () => {
    render(<Button fullWidth>Click</Button>);
    const button = screen.getByRole("button", { name: "Click" });
    expect(button).toBeInTheDocument();
    expect(button).toMatchSnapshot();
  });
});
