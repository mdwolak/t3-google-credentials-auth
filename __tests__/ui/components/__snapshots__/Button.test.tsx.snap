// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Primary disabled 1`] = `
<button
  class="rounded-md font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60 px-4 py-3 text-md bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500"
  disabled=""
  type="button"
>
  Click
</button>
`;

exports[`Button Primary icon only 1`] = `
<button
  class="inline-flex items-center rounded-md font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60 p-3.5 bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500"
  type="button"
>
  <svg
    aria-hidden="true"
    class="h-5 w-5"
    fill="none"
    stroke="currentColor"
    stroke-width="1.5"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 4.5v15m7.5-7.5h-15"
      stroke-linecap="round"
      stroke-linejoin="round"
    />
  </svg>
</button>
`;

exports[`Button Primary loading 1`] = `
<button
  class="inline-flex items-center rounded-md font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60 px-4 py-3 text-md bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500"
  disabled=""
  type="button"
>
  <svg
    class="animate-spin mr-2 h-5 w-5"
    data-testid="spinner"
    fill="none"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      class="opacity-25"
      cx="12"
      cy="12"
      r="10"
      stroke="currentColor"
      stroke-width="4"
    />
    <path
      class="opacity-75"
      d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
      fill="currentColor"
    />
  </svg>
  <span
    class="sr-only"
  >
    Loading
  </span>
  <span>
    Processing...
  </span>
</button>
`;

exports[`Button Primary text 1`] = `
<button
  class="rounded-md font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60 px-4 py-3 text-md bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500"
  type="button"
>
  Click
</button>
`;

exports[`Button Primary text and icon 1`] = `
<button
  class="inline-flex items-center rounded-md font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60 px-4 py-3 text-md bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500"
  type="button"
>
  <svg
    aria-hidden="true"
    class="mr-2 h-5 w-5"
    fill="none"
    stroke="currentColor"
    stroke-width="1.5"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M12 4.5v15m7.5-7.5h-15"
      stroke-linecap="round"
      stroke-linejoin="round"
    />
  </svg>
  Click
</button>
`;

exports[`secondary fullWidth 1`] = `
<button
  class="flex w-full justify-center rounded-md font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60 px-4 py-3 text-md bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500"
  type="button"
>
  Click
</button>
`;

exports[`secondary rounded 1`] = `
<button
  class="rounded-full font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60 px-4 py-3 text-md bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500"
  type="button"
>
  Click
</button>
`;

exports[`secondary text 1`] = `
<button
  class="rounded-md font-medium shadow-sm transition-colors disabled:cursor-not-allowed disabled:opacity-60 px-4 py-3 text-md bg-indigo-600 text-white focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600 enabled:hover:bg-indigo-500"
  type="button"
>
  Click
</button>
`;
